File: backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class RuleRuntimeMetric(BaseModel):
    rule_id: str
    rule_name: Optional[str] = None
    hit_count: int = 0
    total_execution_time_ms: float = 0.0
    avg_execution_time_ms: float = 0.0
    last_fired_at: Optional[datetime] = None


class RuleExplainabilityEntry(BaseModel):
    rule_id: str
    explanation: str
    matched_conditions: List[str] = Field(default_factory=list)
    missing_conditions: List[str] = Field(default_factory=list)
    sample_fact: Dict[str, Any] = Field(default_factory=dict)
    timestamp: datetime


class RuntimeAnalyticsSummary(BaseModel):
    total_rules: int
    triggered_rules: int
    coverage_pct: float
    rules_never_fired_count: int
    events_processed: int
    rules_fired: int
    avg_processing_time_ms: float


class TopRulesResponse(BaseModel):
    top_hot_rules: List[RuleRuntimeMetric] = Field(default_factory=list)
    cold_rules: List[RuleRuntimeMetric] = Field(default_factory=list)


class RuleRuntimeDetail(BaseModel):
    metric: RuleRuntimeMetric
    recent_explanations: List[RuleExplainabilityEntry] = Field(default_factory=list)


class AnalyticsCoverageResponse(BaseModel):
    summary: RuntimeAnalyticsSummary
    triggered_rule_ids: List[str] = Field(default_factory=list)
    never_fired_rule_ids: List[str] = Field(default_factory=list)


class RuntimeAnalyticsResponse(BaseModel):
    summary: RuntimeAnalyticsSummary
    top_hot_rules: List[RuleRuntimeMetric] = Field(default_factory=list)
    cold_rules: List[RuleRuntimeMetric] = Field(default_factory=list)
    recent_explanations: List[RuleExplainabilityEntry] = Field(default_factory=list)

File: frontend/src/types/analytics.ts
export interface RuntimeSummary {
    coverage_pct: number;
    triggered_rules: number;
    total_rules: number;
    rules_never_fired_count: number;
    events_processed: number;
    rules_fired: number;
    avg_processing_time_ms: number;
}

export interface TopRule {
    id: number;
    name: string;
    hit_count: number;
    avg_exec_time_ms: number;
    last_fired?: string | null;
}

export interface RuntimeAnalytics {
    summary: RuntimeSummary;
    top_hot_rules?: TopRule[];
    cold_rules?: TopRule[];
    recent_explanations?: ExplanationItem[];
}

export interface TopRulesResponse {
    top_hot_rules: TopRule[];
    cold_rules: TopRule[];
}

export interface ExplanationItem {
    rule_id: number;
    rule_name: string;
    event_id?: string;
    explanation: string;
    created_at: string;
}

export interface ExplanationsResponse {
    items: ExplanationItem[];
}

export interface GraphSummary {
    total_rules: number;
    filtered_rules: number;
    pair_count: number;
    isolated_rules: ConnectedRule[];
    most_connected_rules: ConnectedRule[];
    top_shared_fields: SharedField[];
    available_groups: string[];
}

export interface ConnectedRule {
    name: string;
    group?: string;
    connections: number;
    field_count: number;
}

export interface SharedField {
    field: string;
    rule_count: number;
    pair_count: number;
}
